crates/sp1/src/host.rs

```diff
@@ -0,0 +1,236 @@
+            let prover = NetworkProverV2::new();
+
+            // Request for proof from Succinct
+            let request_id =
+                block_on(prover.request_proof(self.elf, stdin, ProofMode::Groth16, None))?;
+            // Save pending request id to db
+            self.ledger_db
+                .add_pending_proving_session(request_id.clone())?;
+
+            self.wait_succinct_proof(&prover, request_id)
+        } else {
+            client.prove(&self.proving_key, stdin).groth16().run()
+        }
+    }
+}
+
+impl ZkvmHost for SP1Host {
+    type Guest = SP1Guest;
+
+    fn add_hint<T: BorshSerialize>(&mut self, item: T) {
+        let buf = borsh::to_vec(&item).expect("Borsh hint serialization cannot fail");
+        info!("Added hint to guest with size {}", buf.len());
+
+        // write buf
+        self.input_buf.extend_from_slice(&buf);
+    }
+
+    fn simulate_with_hints(&mut self) -> Self::Guest {
+        unimplemented!("Simulate is not implemented for SP1")
+    }
+
+    fn run(&mut self, with_proof: bool) -> Result<Proof, anyhow::Error> {
+        let stdin = self.collect_input_buf();
+
+        if with_proof {
+            let proof_with_public_values = self.generate_proof(stdin)?;
+            info!("Successfully generated proof");
+
+            self.client
+                .verify(&proof_with_public_values, &self.verifying_key)?;
+            info!("Successfully verified the proof");
+
+            let data = bincode::serialize(&proof_with_public_values)
+                .expect("SP1 zk proof serialization must not fail");
+            Ok(Proof::Full(data))
+        } else {
+            let (public_values, report) = self.client.execute(self.elf, stdin).run()?;
+            info!("Number of cycles: {}", report.total_instruction_count());
+
+            let data = bincode::serialize(&public_values)
+                .expect("SP1 zk public values serialization must not fail");
+            Ok(Proof::PublicInput(data))
+        }
+    }
+
+    fn extract_output<Da: sov_rollup_interface::da::DaSpec, Root: BorshDeserialize>(
+        proof: &Proof,
+    ) -> Result<sov_rollup_interface::zk::StateTransition<Da, Root>, Self::Error> {
+        let public_values = match proof {
+            Proof::PublicInput(data) => {
+                let public_values: SP1PublicValues = bincode::deserialize(data)?;
+                public_values
+            }
+            Proof::Full(data) => {
+                let proof: SP1ProofWithPublicValues = bincode::deserialize(data)?;
+                proof.public_values
+            }
+        };
+
+        Ok(BorshDeserialize::try_from_slice(public_values.as_slice())?)
+    }
+
+    fn recover_proving_sessions(&self) -> Result<Vec<Proof>, anyhow::Error> {
+        // We can only recover if prover is configured to be Succinct
+        if !self.is_succinct_prover() {
+            return Ok(vec![]);
+        }
+
+        let request_ids = self.ledger_db.get_pending_proving_sessions()?;
```

Comment:
Probably we have to separate risc0 and sp1 sessions

Reply:
the fact that we store sessions as Vec<u8> helped a lot in using both type of sessions without needing to add another ledgerdb functionality. IMO this is fine at least for now since we technically don't ever use SP1 in the codebase.

Reply:
We can store (kind, session) in one get_pending_proving_sessions. Otherwise it will be painful to debug in the future

Reply:
You are correct. But this is only going to be a problem when we are switching from risc0 to sp1. We will never have pending sessions from both, except, potentially when we restart the prover through deployment, and right at the restart, there was a risc0 proof happening, and with that deployment we switched to sp1.

I think this should be handled within the pr where we actually switch sp1, if we ever.

Reply:
Yea i agree. We know that we won't use both sessions simultaneously so we don't have to support both in storage.

As for how we handle storage related to risc0 when we switch to SP1 (if ever), would be in a ticket i am working on:
https://github.com/chainwayxyz/citrea/issues/1338